firmware/main.c

```diff
@@ -99,13 +105,18 @@ uchar usbFunctionSetup(uchar data[8]) {
 
 		/*
 		 * Auto sck probing is done here vs in FUNC_ENABLEPROG
-		 * So that host applications that use FUNC_TRANSMIT instead to
+		 * So that host applications that use FUNC_TRANSMIT to do
 		 * their own raw ISP port access instead of letting the USBasp device
-		 * do it form them, will also benefit from the auto sck probing.
+		 * do it for them will also benefit from the auto sck probing.
 		 */
 
 		if (prog_sck == USBASP_ISP_SCK_AUTO) {
-			prog_sck = autosck(USBASP_ISP_SCK_1500);
+			/*
```

Comment:
Though introducing the 4KHz speed fixes 32KHz, it (sometimes) breaks 16KHz. When flashing an attiny85 I saw the first connect selecting 1KHz, but the second one selecting 4KHz, which is on the edge of the spec and failed...
This is what I meant with the auto_sck getting fuzzy... I think it is unavoidable that some target cpu speeds are not auto probed, so we should not try too hard either. Since 16KHz is selectable through the internal osc, I think it is more important than 32KHz. Therefore I have a preference for the speed array without 4KHz.

Also, if you have to fall back to a sow clock, we should fall back to the slowest we have: 1KHz (for what it is worth, next attempts will fail anyway...)

Also, jp3 has always selected 8KHz, many people know this, it is documented in many places, I have a slight preference to keep it.

That said, I agree with most of the measures taken in this commit (unresetting before trying each speed, retrying the selected speed).